### ISSUE TITLE
named-refused rule seems not to be working

### ISSUE THREAD

**Issue by ksingvo** [fail2ban/fail2ban#2899]
### Environment:

- Fail2Ban version (including any possible distribution suffixes): 0.11.2-lp152.118.1
- OS, including release name/version: openSUSE Leap 15.2
- [x] Fail2Ban installed via OS/distribution mechanisms
- [x] You have not applied any additional foreign patches to the codebase
- [x] Some customizations were done to the configuration (provide details below is so)

### The issue:

Trying to ban "query (cache) ... denied" requests for named, but the rules fail with " Prefregex not matched"

First I worked with the outdated version 0.10.4 of fail2ban shipped with openSUSE distribution, then I switched to the version 0.11.2 avail in the openSUSE BuilService  (Security/fail2ban):
[https://build.opensuse.org/package/show/security/fail2ban](https://build.opensuse.org/package/show/security/fail2ban)

#### Any additional information

`fail2ban-regex "syslog line" /etc/fail2ban/filter.d/named-refused.conf` shows to my surprise a match

#### Any customizations done to /etc/fail2ban/ configuration
`jail.local`
```
[DEFAULT]
usedns=yes
# "bantime" is the number of seconds that a host is banned.
bantime=3w

# A host is banned if it has generated "maxretry" during the last "findtime"
# seconds.
findtime=15d

# "maxretry" is the number of failures before a host get banned.
maxretry=10

backend=systemd

##########

[sshd]
port=ssh,sftp
enabled=true

[pam-generic]
enabled=true

[postfix]
enabled=true

[named-refused]
enabled=true
logpath=/var/log/messages

##########

[named-refused-udp]
enabled=true
port=domain,953
protocol=udp
filter=named-refused
action=iptables-multiport[name=Named, port="domain,953", protocol=udp]
logpath=/var/log/messages
findtime=7d
bantime=3w
maxretry=3

##########

[named-refused-tcp]
enabled   = true
port      = domain,953
protocol  = tcp
filter    = named-refused
action    = iptables-multiport[name=Named, port="domain,953", protocol=tcp]
logpath   = /var/log/messages
findtime  = 43200
bantime   = 604800
maxretry  = 1

##########
```

`filter.d/named-refused.conf`
```
# Fail2Ban filter file for named (bind9).
#

# This filter blocks attacks against named (bind9) however it requires special
# configuration on bind.
#
# By default, logging is off with bind9 installation.
#
# You will need something like this in your named.conf to provide proper logging.
#
# logging {
#     channel security_file {
#         file "/var/log/named/security.log" versions 3 size 30m;
#         severity dynamic;
#         print-time yes;
#     };
#     category security {
#         security_file;
#     };
# };

[Definition]

# Daemon name
_daemon=named

# Shortcuts for easier comprehension of the failregex

__pid_re=(?:\[\d+\])
__daemon_re=\(?%(_daemon)s(?:\(\S+\))?\)?:?
__daemon_combs_re=(?:%(__pid_re)s?:\s+%(__daemon_re)s|%(__daemon_re)s%(__pid_re)s?:)

#       hostname       daemon_id         spaces
# this can be optional (for instance if we match named native log files)
__line_prefix=(?:\s\S+ %(__daemon_combs_re)s\s+)?

prefregex = ^%(__line_prefix)s(?: error:)?\s*client(?: @\S*)? <HOST>#\S+(?: \([\S.]+\))?: <F-CONTENT>.+</F-CONTENT>\s(?:denied|\(NOTAUTH\))\s*$

failregex = ^(?:view (?:internal|external): )?query(?: \(cache\))?
            ^zone transfer
            ^bad zone transfer request: '\S+/IN': non-authoritative zone

ignoreregex =

# DEV Notes:
# Trying to generalize the
#          structure which is general to capture general patterns in log
#          lines to cover different configurations/distributions
#          
# Author: Yaroslav Halchenko
```

#### Relevant parts of /var/log/fail2ban.log file:
_preferably obtained while running fail2ban with `loglevel = 4`_

```
2020-12-29 17:16:41,260 fail2ban.filtersystemd  [19574]: HEAVY   [pam-generic] Read systemd journal entry: 2020-12-29T17:16:40.954246 atom named[1806]: client @0x7fb13400eec0 78.47.191.45#61977 (.): query (cache) './ANY/IN' denied
2020-12-29 17:16:41,261 fail2ban.filtersystemd  [19574]: HEAVY   [named-refused-tcp] Read systemd journal entry: 2020-12-29T17:16:40.954246 atom named[1806]: client @0x7fb13400eec0 78.47.191.45#61977 (.): query (cache) './ANY/IN' denied
2020-12-29 17:16:41,250 fail2ban.filtersystemd  [19574]: HEAVY   [named-refused-udp] Read systemd journal entry: 2020-12-29T17:16:40.956048 atom named[1806]: client @0x7fb13400eec0 78.47.191.45#61977 (.): query (cache) './ANY/IN' denied
2020-12-29 17:16:41,262 fail2ban.filtersystemd  [19574]: HEAVY   [named-refused] Read systemd journal entry: 2020-12-29T17:16:40.956048 atom named[1806]: client @0x7fb13400eec0 78.47.191.45#61977 (.): query (cache) './ANY/IN' denied
2020-12-29 17:16:41,264 fail2ban.filter         [19574]: TRACE   Working on line ('', '2020-12-29T17:16:40.956048', "atom named[1806]: client @0x7fb13400eec0 78.47.191.45#61977 (.): query (cache) './ANY/IN' denied")
2020-12-29 17:16:41,263 fail2ban.filter         [19574]: TRACE   Working on line ('', '2020-12-29T17:16:40.954246', "atom named[1806]: client @0x7fb13400eec0 78.47.191.45#61977 (.): query (cache) './ANY/IN' denied")
2020-12-29 17:16:41,265 fail2ban.filter         [19574]: HEAVY   Looking for match of [('', '2020-12-29T17:16:40.954246', "atom named[1806]: client @0x7fb13400eec0 78.47.191.45#61977 (.): query (cache) './ANY/IN' denied")]
2020-12-29 17:16:41,264 fail2ban.filter         [19574]: HEAVY   Looking for match of [('', '2020-12-29T17:16:40.956048', "atom named[1806]: client @0x7fb13400eec0 78.47.191.45#61977 (.): query (cache) './ANY/IN' denied")]
2020-12-29 17:16:41,263 fail2ban.filter         [19574]: TRACE   Working on line ('', '2020-12-29T17:16:40.954246', "atom named[1806]: client @0x7fb13400eec0 78.47.191.45#61977 (.): query (cache) './ANY/IN' denied")
2020-12-29 17:16:41,265 fail2ban.filter         [19574]: HEAVY     Looking for prefregex '^\\s*(?:\\S+\\s+)?(?:\\S+(?:\\[\\d+\\])?:?\\s+)?(?:kernel:\\s?\\[ *\\d+\\.\\d+\\]:?\\s+)?\\(?pam_unix(?:\\(\\S+\\))?\\)?:?\\s+authentication failure;(?:\\s+(?:(?:logname|e?uid)=\\S*)){0,3} tty=\\S* (?P<content>.+)$'
2020-12-29 17:16:41,266 fail2ban.filter         [19574]: HEAVY     Prefregex not matched
2020-12-29 17:16:41,265 fail2ban.filter         [19574]: HEAVY     Looking for prefregex '^(?:\\s\\S+ (?:(?:\\[\\d+\\])?:\\s+\\(?named(?:\\(\\S+\\))?\\)?:?|\\(?named(?:\\(\\S+\\))?\\)?:?(?:\\[\\d+\\])?:)\\s+)?(?: error:)?\\s*client(?: @\\S*)? (?:\\[?(?:(?:::f{4,6}:)?(?P<ip4>(?:\\d{1,3}\\.){3}\\d{1,3})|(?P<ip6>(?:[0-9a-fA-F]{1,4}::?|::){1,7}(?:[0-9a-fA-F]{1,4}|(?<=:):)))\\]?|(?P<dns>[\\w\\-.^_]*\\w))#\\S+(?: \\([\\S.]+\\))?: (?P<content>.+)\\s(?:denied|\\(NOTAUTH\\))\\s*$'
2020-12-29 17:16:41,268 fail2ban.filter         [19574]: HEAVY     Prefregex not matched
2020-12-29 17:16:41,266 fail2ban.filter         [19574]: HEAVY   Looking for match of [('', '2020-12-29T17:16:40.954246', "atom named[1806]: client @0x7fb13400eec0 78.47.191.45#61977 (.): query (cache) './ANY/IN' denied")]
2020-12-29 17:16:41,267 fail2ban.filtersystemd  [19574]: HEAVY   [pam-generic] Read systemd journal entry: 2020-12-29T17:16:40.956048 atom named[1806]: client @0x7fb13400eec0 78.47.191.45#61977 (.): query (cache) './ANY/IN' denied
2020-12-29 17:16:41,263 fail2ban.filter         [19574]: TRACE   Working on line ('', '2020-12-29T17:16:40.956048', "atom named[1806]: client @0x7fb13400eec0 78.47.191.45#61977 (.): query (cache) './ANY/IN' denied")
2020-12-29 17:16:41,269 fail2ban.filter         [19574]: HEAVY   Looking for match of [('', '2020-12-29T17:16:40.956048', "atom named[1806]: client @0x7fb13400eec0 78.47.191.45#61977 (.): query (cache) './ANY/IN' denied")]
2020-12-29 17:16:41,269 fail2ban.filter         [19574]: TRACE   Working on line ('', '2020-12-29T17:16:40.956048', "atom named[1806]: client @0x7fb13400eec0 78.47.191.45#61977 (.): query (cache) './ANY/IN' denied")
2020-12-29 17:16:41,268 fail2ban.filter         [19574]: HEAVY     Looking for prefregex '^(?:\\s\\S+ (?:(?:\\[\\d+\\])?:\\s+\\(?named(?:\\(\\S+\\))?\\)?:?|\\(?named(?:\\(\\S+\\))?\\)?:?(?:\\[\\d+\\])?:)\\s+)?(?: error:)?\\s*client(?: @\\S*)? (?:\\[?(?:(?:::f{4,6}:)?(?P<ip4>(?:\\d{1,3}\\.){3}\\d{1,3})|(?P<ip6>(?:[0-9a-fA-F]{1,4}::?|::){1,7}(?:[0-9a-fA-F]{1,4}|(?<=:):)))\\]?|(?P<dns>[\\w\\-.^_]*\\w))#\\S+(?: \\([\\S.]+\\))?: (?P<content>.+)\\s(?:denied|\\(NOTAUTH\\))\\s*$'
2020-12-29 17:16:41,271 fail2ban.filter         [19574]: HEAVY     Prefregex not matched
2020-12-29 17:16:41,270 fail2ban.filter         [19574]: HEAVY   Looking for match of [('', '2020-12-29T17:16:40.956048', "atom named[1806]: client @0x7fb13400eec0 78.47.191.45#61977 (.): query (cache) './ANY/IN' denied")]
2020-12-29 17:16:41,270 fail2ban.filter         [19574]: HEAVY     Looking for prefregex '^(?:\\s\\S+ (?:(?:\\[\\d+\\])?:\\s+\\(?named(?:\\(\\S+\\))?\\)?:?|\\(?named(?:\\(\\S+\\))?\\)?:?(?:\\[\\d+\\])?:)\\s+)?(?: error:)?\\s*client(?: @\\S*)? (?:\\[?(?:(?:::f{4,6}:)?(?P<ip4>(?:\\d{1,3}\\.){3}\\d{1,3})|(?P<ip6>(?:[0-9a-fA-F]{1,4}::?|::){1,7}(?:[0-9a-fA-F]{1,4}|(?<=:):)))\\]?|(?P<dns>[\\w\\-.^_]*\\w))#\\S+(?: \\([\\S.]+\\))?: (?P<content>.+)\\s(?:denied|\\(NOTAUTH\\))\\s*$'
2020-12-29 17:16:41,272 fail2ban.filter         [19574]: HEAVY     Prefregex not matched
2020-12-29 17:16:41,272 fail2ban.filter         [19574]: HEAVY     Looking for prefregex '^\\s*(?:\\S+\\s+)?(?:\\S+(?:\\[\\d+\\])?:?\\s+)?(?:kernel:\\s?\\[ *\\d+\\.\\d+\\]:?\\s+)?\\(?pam_unix(?:\\(\\S+\\))?\\)?:?\\s+authentication failure;(?:\\s+(?:(?:logname|e?uid)=\\S*)){0,3} tty=\\S* (?P<content>.+)$'
2020-12-29 17:16:41,272 fail2ban.filtersystemd  [19574]: HEAVY   [named-refused-tcp] Read systemd journal entry: 2020-12-29T17:16:40.956048 atom named[1806]: client @0x7fb13400eec0 78.47.191.45#61977 (.): query (cache) './ANY/IN' denied
2020-12-29 17:16:41,273 fail2ban.filter         [19574]: HEAVY     Prefregex not matched
2020-12-29 17:16:41,273 fail2ban.filter         [19574]: TRACE   Working on line ('', '2020-12-29T17:16:40.956048', "atom named[1806]: client @0x7fb13400eec0 78.47.191.45#61977 (.): query (cache) './ANY/IN' denied")
2020-12-29 17:16:41,274 fail2ban.filter         [19574]: HEAVY   Looking for match of [('', '2020-12-29T17:16:40.956048', "atom named[1806]: client @0x7fb13400eec0 78.47.191.45#61977 (.): query (cache) './ANY/IN' denied")]
2020-12-29 17:16:41,274 fail2ban.filter         [19574]: HEAVY     Looking for prefregex '^(?:\\s\\S+ (?:(?:\\[\\d+\\])?:\\s+\\(?named(?:\\(\\S+\\))?\\)?:?|\\(?named(?:\\(\\S+\\))?\\)?:?(?:\\[\\d+\\])?:)\\s+)?(?: error:)?\\s*client(?: @\\S*)? (?:\\[?(?:(?:::f{4,6}:)?(?P<ip4>(?:\\d{1,3}\\.){3}\\d{1,3})|(?P<ip6>(?:[0-9a-fA-F]{1,4}::?|::){1,7}(?:[0-9a-fA-F]{1,4}|(?<=:):)))\\]?|(?P<dns>[\\w\\-.^_]*\\w))#\\S+(?: \\([\\S.]+\\))?: (?P<content>.+)\\s(?:denied|\\(NOTAUTH\\))\\s*$'
2020-12-29 17:16:41,275 fail2ban.filter         [19574]: HEAVY     Prefregex not matched
2020-12-29 17:16:42,255 fail2ban.actions        [19574]: HEAVY   Actions: check-unban 20, bancnt 0, max: 20
2020-12-29 17:16:42,256 fail2ban.actions        [19574]: HEAVY   Actions: wait for pending tickets 2 (default 2)
2020-12-29 17:16:42,718 fail2ban.actions        [19574]: HEAVY   Actions: check-unban 20, bancnt 0, max: 20
```

#### Relevant lines from monitored log files in question:

`systemd` output, not a logfile:
```
2020-12-29T17:16:40.954246 atom named[1806]: client @0x7fb13400eec0 78.47.191.45#61977 (.): query (cache) './ANY/IN' denied
```

**Comment by sebres** [comment]
Despite you defined `logpath` in the jails it will be ignored (due to default setting `backend=systemd` or some platform includes), so your fail2ban is monitoring systemd journal for the jails.

And it looks like the filter `named-refused` is indeed not fully systemd journal compatible. If you'd take a look at this line:
https://github.com/fail2ban/fail2ban/blob/91923b5c072fc6209a9e9aff06f91b7d896ec63f/config/filter.d/named-refused.conf#L35
you'll probably notice that a space at begin of line is mandatory (due to first `\s`). 
But it is missing if line is coming from systemd journal (the message arrives without timestamp at all).

I'll try to fix it soon... Unless it is not released, you could redefine the parameter `__line_prefix` for filter in your `jail.local` like here:
```ini
[named-refused]
filter = %(known/filter)s[__line_prefix="\s*\S+\s+named\[\d+\]:"]
action = %(banaction)s[name=%(__name__)s-tcp, port="%(port)s", protocol="tcp", chain="%(chain)s", actname=%(banaction)s-tcp]
         %(banaction)s[name=%(__name__)s-udp, port="%(port)s", protocol="udp", chain="%(chain)s", actname=%(banaction)s-udp]
enabled = true
```

To test within fail2ban-regex use:
```bash
fail2ban-regex systemd-journal 'named-refused[__line_prefix="\s*\S+\s+named\[\d+\]:"]'
```

By the way, you don't need 2 separate jails for this - just specify 2 actions as in example above in the single jail.

**Comment by gelinger777** [comment]
the latest rule config in repo do not work with  following lines

```
03-Jan-2021 17:01:59.481 client @0x7f6efc1aeb30 81.84.147.225#23510 (.): query (cache) './ANY/IN' denied
03-Jan-2021 17:01:59.481 client @0x7f6efc191c10 81.84.147.225#23510 (.): query (cache) './ANY/IN' denied
03-Jan-2021 17:01:59.481 client @0x7f6efc1cba50 81.84.147.225#23510 (.): query (cache) './ANY/IN' denied
03-Jan-2021 17:01:59.481 client @0x7f6efc1bd2c0 81.84.147.225#23510 (.): query (cache) './ANY/IN' denied
03-Jan-2021 17:01:59.481 client @0x7f6efc183480 81.84.147.225#23510 (.): query (cache) './ANY/IN' denied
03-Jan-2021 17:02:05.798 client @0x7f6efc183480 99.41.233.140#35164 (.): query (cache) './ANY/IN' denied
03-Jan-2021 17:02:05.799 client @0x7f6efc1bd2c0 99.41.233.140#35164 (.): query (cache) './ANY/IN' denied
03-Jan-2021 17:02:05.799 client @0x7f6efc191c10 99.41.233.140#35164 (.): query (cache) './ANY/IN' denied
03-Jan-2021 17:02:05.799 client @0x7f6efc174cf0 99.41.233.140#35164 (.): query (cache) './ANY/IN' denied
03-Jan-2021 17:02:05.799 client @0x7f6efc1cba50 99.41.233.140#35164 (.): query (cache) './ANY/IN' denied
03-Jan-2021 17:02:35.237 client @0x7f6efc1cba50 81.84.147.225#58022 (.): query (cache) './ANY/IN' denied
03-Jan-2021 17:02:35.237 client @0x7f6efc174cf0 81.84.147.225#58022 (.): query (cache) './ANY/IN' denied
03-Jan-2021 17:02:35.237 client @0x7f6efc183480 81.84.147.225#58022 (.): query (cache) './ANY/IN' denied
03-Jan-2021 17:02:35.237 client @0x7f6efc191c10 81.84.147.225#58022 (.): query (cache) './ANY/IN' denied
03-Jan-2021 17:02:35.238 client @0x7f6efc1bd2c0 81.84.147.225#58022 (.): query (cache) './ANY/IN' denied
03-Jan-2021 17:02:55.798 client @0x7f6efc191c10 81.84.147.225#13626 (.): query (cache) './ANY/IN' denied
03-Jan-2021 17:02:55.798 client @0x7f6efc1bd2c0 81.84.147.225#13626 (.): query (cache) './ANY/IN' denied
03-Jan-2021 17:02:55.798 client @0x7f6efc183480 81.84.147.225#13626 (.): query (cache) './ANY/IN' denied
03-Jan-2021 17:02:55.798 client @0x7f6efc174cf0 81.84.147.225#13626 (.): query (cache) './ANY/IN' denied

```

how to fix it?

**Comment by sebres** [comment]
> the latest rule config in repo do not work with following lines

How so? (it does well)... What do you mean as latest?

#### PoC:
```
fail2ban-regex /tmp/log named-refused

Running tests
=============

Use   failregex filter file : named-refused, basedir: /etc/fail2ban/
Use         log file : /tmp/log
Use         encoding : UTF-8


Results
=======

Prefregex: 19 total
|  ^(?:\s*\S+ (?:(?:\[\d+\])?:\s+\(?named(?:\(\S+\))?\)?:?|\(?named(?:\(\S+\))?\)?:?(?:\[\d+\])?:)\s+)?(?: error:)?\s*client(?: @\S*)? (?:\[?(?:(?:::f{4,6}:)?(?P<ip4>(?:\d{1,3}\.){3}\d{1,3})|(?P<ip6>(?:[0-9a-fA-F]{1,4}::?|::){1,7}(?:[0-9a-fA-F]{1,4}|(?<=:):)))\]?|(?P<dns>[\w\-.^_]*\w))#\S+(?: \([\S.]+\))?: (?P<content>.+)\s(?:denied|\(NOTAUTH\))\s*$
`-

Failregex: 19 total
|-  #) [# of hits] regular expression
|   1) [19] ^(?:view (?:internal|external): )?query(?: \(cache\))?
`-

Ignoreregex: 0 total

Date template hits:
|- [# of hits] date format
|  [19] {^LN-BEG}Day(?P<_sep>[-/])MON(?P=_sep)ExYear[ :]?24hour:Minute:Second(?:\.Microseconds)?(?: Zone offset)?
`-

Lines: 19 lines, 0 ignored, 19 matched, 0 missed
[processed in 0.01 sec]
```

The issue was about troubles by the filter together with messages from systemd journal (systemd backend only) and it is closed in 73b39e08947b1858d7574085c3f74318e101e0de
If you have the same thing, update your config like in this patch and it should work.

**Comment by gelinger777** [comment]
I have updated the tools and regex file and it has started to work. I don't know why it was not...  Thanks
